File: src/python/harness_engine/envelope.py
"""Data contract exchanged between the driver (agent) and the state machine.

The model returns this envelope as JSON; the engine dispatches by `value`.

There is no tokens field: the typical driver is an LLM with no access to its own
request's `usage`, so any self-reported count would be confabulated. The cost ceiling
only uses measures the engine attests itself (steps and instruction chars — see
`task_registry`); real tokens live in the caller's billing metadata.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass

from harness_engine.context_policy import ContextUsage


class EnvelopeType:
    """Protocol signals carried in `Envelope.type`."""

    TEXT = "text"
    TOOL = "tool"
    COMMAND = "command"
    ERROR = "error"


@dataclass(frozen=True)
class Envelope:
    type: str
    value: str
    # Tuple (not list): immutable and comparable by value for free — C# needs manual
    # Equals/GetHashCode because arrays compare by reference; not the case here.
    args: tuple[str, ...] = ()
    context: dict[str, str] | None = None
    context_usage: ContextUsage | None = None

    def to_json(self) -> str:
        data: dict[str, object] = {
            "type": self.type,
            "value": self.value,
            "args": list(self.args),
        }
        if self.context is not None:
            data["context"] = self.context
        if self.context_usage is not None:
            data["contextUsage"] = self.context_usage.to_dict()
        return json.dumps(data, separators=(",", ":"))

    @staticmethod
    def parse(value: str) -> "Envelope | None":
        """Tolerant parse: accepts markdown fences and surrounding text around the JSON object."""
        return Envelope._try_parse(value)

    @staticmethod
    def _try_parse(value: str) -> "Envelope | None":
        try:
            if value is None or not value.strip():
                raise ValueError("The JSON envelope cannot be null or empty.")

            root = json.loads(Envelope._sanitize(value))
            if not isinstance(root, dict):
                raise ValueError("The envelope payload must be a JSON object.")

            type_ = root.get("type") or ""
            envelope_value = root.get("value") or ""
            if not isinstance(type_, str) or not isinstance(envelope_value, str):
                raise TypeError("'type' and 'value' must be strings.")

            args_raw = root.get("args")
            args: tuple[str, ...] = ()
            if isinstance(args_raw, list):
                collected: list[str] = []
                for item in args_raw:
                    if not isinstance(item, str):
                        raise TypeError("each item in 'args' must be a string.")
                    if item.strip():
                        collected.append(item)
                args = tuple(collected)

            context_raw = root.get("context")
            context: dict[str, str] | None = None
            if isinstance(context_raw, dict):
                context = {}
                for key, val in context_raw.items():
                    if not isinstance(val, str):
                        raise TypeError("each value in 'context' must be a string.")
                    context[str(key)] = val

            context_usage = ContextUsage.from_dict(root.get("contextUsage"))

            return Envelope(type_, envelope_value, args, context, context_usage)
        except Exception as ex:
            # Diagnostics go to stderr — stdout is the harness transport channel (the
            # driver reads stdout as the next instruction) and must not be polluted.
            print(ex, file=sys.stderr)
            return None

    # Models frequently wrap the JSON in markdown fences (```json … ```) or add
    # surrounding text. Normalize to the raw JSON object before parsing.
    @staticmethod
    def _sanitize(value: str) -> str:
        v = value.strip()

        if v.startswith("```"):
            first_newline = v.find("\n")
            if first_newline >= 0:
                v = v[first_newline + 1 :]

            closing_fence = v.rfind("```")
            if closing_fence >= 0:
                v = v[:closing_fence]

            v = v.strip()

        start = v.find("{")
        end = v.rfind("}")
        if start >= 0 and end > start:
            v = v[start : end + 1]

        return v

File: src/go/harnessengine/envelope.go
// Package harnessengine is the domain-agnostic core of the harness — the Go port of
// Harness.Engine (.NET), harness_engine (Python/Rust). It owns entry/dispatch, prompt
// formatting, and persistence/telemetry; a flow (e.g. flowsdevelopment) owns domain policy.
package harnessengine

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// EnvelopeType holds the protocol signals carried in Envelope.Type.
var EnvelopeType = struct {
	Text    string
	Tool    string
	Command string
	Error   string
}{
	Text:    "text",
	Tool:    "tool",
	Command: "command",
	Error:   "error",
}

// Envelope is the data contract exchanged between the driver (agent) and the state
// machine. The model returns this envelope as JSON; the engine dispatches by Value.
//
// There is no tokens field: the typical driver is an LLM with no access to its own
// request's usage, so any self-reported count would be confabulated. The cost ceiling
// only uses measures the engine attests itself (steps and instruction chars — see
// TaskRegistry); real tokens live in the caller's billing metadata.
type Envelope struct {
	Type         string            `json:"type"`
	Value        string            `json:"value"`
	Args         []string          `json:"args"`
	Context      map[string]string `json:"context,omitempty"`
	ContextUsage *ContextUsage     `json:"contextUsage,omitempty"`
}

// NewEnvelope builds an envelope with no context (args defaults to an empty, non-nil slice).
func NewEnvelope(envelopeType, value string, args []string) Envelope {
	if args == nil {
		args = []string{}
	}
	return Envelope{Type: envelopeType, Value: value, Args: args}
}

// ToJSON serializes the envelope compactly — the same wire format as .NET/Python/Rust.
func (e Envelope) ToJSON() string {
	b, err := json.Marshal(e)
	if err != nil {
		// Envelope only carries strings/maps of strings; marshaling never fails in practice.
		return "{}"
	}
	return string(b)
}

// ParseEnvelope is a tolerant parse: accepts markdown fences and surrounding text around
// the JSON object.
func ParseEnvelope(value string) *Envelope {
	envelope, err := tryParseEnvelope(value)
	if err != nil {
		// Diagnostics go to stderr — stdout is the harness transport channel (the driver
		// reads stdout as the next instruction) and must not be polluted.
		fmt.Fprintln(os.Stderr, err)
		return nil
	}
	return envelope
}

func tryParseEnvelope(value string) (*Envelope, error) {
	if strings.TrimSpace(value) == "" {
		return nil, fmt.Errorf("The JSON envelope cannot be null or empty.")
	}

	sanitized := sanitizeEnvelope(value)

	var root map[string]json.RawMessage
	if err := json.Unmarshal([]byte(sanitized), &root); err != nil {
		return nil, err
	}

	envType, err := stringField(root, "type")
	if err != nil {
		return nil, err
	}
	envValue, err := stringField(root, "value")
	if err != nil {
		return nil, err
	}

	args := []string{}
	if raw, ok := root["args"]; ok {
		var items []json.RawMessage
		if err := json.Unmarshal(raw, &items); err == nil {
			for _, item := range items {
				var s string
				if err := json.Unmarshal(item, &s); err != nil {
					return nil, fmt.Errorf("cada item de 'args' deve ser uma string.")
				}
				if strings.TrimSpace(s) != "" {
					args = append(args, s)
				}
			}
		}
	}

	var context map[string]string
	if raw, ok := root["context"]; ok {
		var obj map[string]json.RawMessage
		if err := json.Unmarshal(raw, &obj); err == nil {
			context = make(map[string]string, len(obj))
			for k, v := range obj {
				var s string
				if err := json.Unmarshal(v, &s); err != nil {
					return nil, fmt.Errorf("cada valor de 'context' deve ser uma string.")
				}
				context[k] = s
			}
		}
	}

	var contextUsage *ContextUsage
	if raw, ok := root["contextUsage"]; ok {
		var usage ContextUsage
		if err := json.Unmarshal(raw, &usage); err == nil {
			contextUsage = &usage
		}
	}

	return &Envelope{Type: envType, Value: envValue, Args: args, Context: context, ContextUsage: contextUsage}, nil
}

// stringField reads an optional string field: absent/null becomes "", any other type is a
// parse error.
func stringField(root map[string]json.RawMessage, key string) (string, error) {
	raw, ok := root[key]
	if !ok {
		return "", nil
	}
	if string(raw) == "null" {
		return "", nil
	}
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		return "", fmt.Errorf("'%s' deve ser uma string.", key)
	}
	return s, nil
}

// sanitizeEnvelope normalizes model output that wraps JSON in markdown fences or adds
// surrounding prose into the raw JSON object.
func sanitizeEnvelope(value string) string {
	v := strings.TrimSpace(value)

	if strings.HasPrefix(v, "```") {
		if idx := strings.IndexByte(v, '\n'); idx >= 0 {
			v = v[idx+1:]
		}
		if idx := strings.LastIndex(v, "```"); idx >= 0 {
			v = v[:idx]
		}
		v = strings.TrimSpace(v)
	}

	start := strings.IndexByte(v, '{')
	end := strings.LastIndexByte(v, '}')
	if start >= 0 && end > start {
		v = v[start : end+1]
	}

	return v
}
